### PR TITLE
Drop database views

### PR DESCRIPTION
**Pull request by jdavcs**
See alternative implementation in #15888 

IMPORTANT: I've dropped [this code](https://github.com/galaxyproject/galaxy/pull/15876/commits/4d26fc86c142d83382681b483b946f7986278813) (joinedload option for job_state_summary in `history_contents` manager). I'm not sure if this breaks something (or negatively affects performance). If this needs to be restored, my suggestion is to go with #15888, since this version does not use a mapped class and so it cannot pull these summary counts as a query load option.

Why drop views:
- A view is a database-level abstraction we simply don't need: we define our abstractions in the code base, not the database. There's no difference in terms of convenience or performance between querying a database view or a SQLAlchemy construct like a Selectable.
- Views add unnecessary complexity for mapping: it's a custom DDL element that requires some setup (see reference implementation on [sqlalchemy's wiki](https://github.com/sqlalchemy/sqlalchemy/wiki/Views)), and, especially, database migrations (see Alembic's [replaceable objects](https://alembic.sqlalchemy.org/en/latest/cookbook.html#replaceable-objects)).

Unlike #15888, this is a more intrusive refactoring of `HistoryDatasetCollectionJobStateSummary`'s implementation. 
1. Here we remove the mapped class and instead use a simple query. I think there is no need for a mapped class: there is no state that the ORM needs to track: it's just one readonly row containing 16 integer values which are aggregate counts. Executing a query seems like a better fit: is simple, and it's much more flexible than dealing with a mapped class. (also, as [per SA's docs](https://docs.sqlalchemy.org/en/20/orm/nonstandard_mappings.html#mapping-a-class-against-arbitrary-subqueries), this is not a recommended approach)
2. I've simplified the query. It returns the same data (I've verified this on main). The previous version used an outer join, which resulted in containing rows that did not add anything to the counts (rows having no associated job contributed neither state counts, nor total count), which is why a `group by` clause was necessary. More importantly, the previous query could only be used in a join - i.e., the class  was ONLY used to define a relationship with a parent class - which added the HDCA.id as the join criteria. Without it the query is ..expensive: cost=20432585 (so using the class as a standalone makes no sense). 
3. I've rearranged the joins to make the logic follow the join path: we select job state, joining along this path: icjja > icj > hdca. Again, there's no difference in what data is produced by the query.
4. If there are no associated jobs, the simplified query returns a tuple of 15 nulls and one zero for total jobs. That's because we don't select the jobs we won't need in the first place, and the `sum` function returns a null if there are no rows to sum over. We fix this in the code by simply assigning zeros to all counts if the total count is zero.

SAME NOTE as in 15888: The part of the query that enumerates the job states (aggregate `sum` columns) is now generated dynamically based on members of `model.Job.states`. This was done as a safeguard against missing a state (assertion added) or listing a non-existent state (e.g. the original view's query had a bug: the state `deleted_new` has been replaced with `deleting`, which may have resulted in incorrect counts.)

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by jdavcs** [comment]
Superseded by #15888.

**Comment by jdavcs** [comment]
Nope, let's try again. I think I found the problem.

**Comment by mvdbeek** [comment]
This looks fine to me, but don't you need to explicitly drop the view in a new migration ?

**Comment by jdavcs** [comment]
@mvdbeek Thank you for looking at these!

I've added the revision for dropping the view (I had it in the PR's initial to-dos but then lost it in the edits.. Thanks!)

I've rewritten the revision script that edited the view (c39f1de47a04) without the use of alembic_util. I think that's all we need. I've tested it - works both ways. I've also dropped the alembic_util dependency in the next commit - I don't think we need it anymore.

File: lib/galaxy/util/__init__.py [hunk 1]
```diff
@@ -1905,3 +1905,11 @@ def to_str(self, **kwd):
...
     def elapsed(self):
         return time.time() - self.begin
+
+
+def enum_values(enum_class):
+    """
+    Return a list of member values of enumeration enum_class.
+    Values are in member definition order.
+    """
+    return [item.value for key, item in enum_class.__members__.items()]
```
Comment (nsoranzo) [inline lib/galaxy/util/__init__.py]:
```suggestion
    return [value.value for value in enum_class.__members__.values()]
```

File: lib/galaxy/model/__init__.py [hunk 2]
```diff
@@ -10555,3 +10614,6 @@ def receive_init(target, args, kwargs):
         if obj:
             add_object_to_object_session(target, obj)
             return  # Once is enough.
+
+
+JobStateSummary = namedtuple("JobStateSummary", enum_values(Job.states) + ["all_jobs"])  # type: ignore[misc]  # Ref https://github.com/python/mypy/issues/848#issuecomment-255237167
```
Comment (nsoranzo) [inline lib/galaxy/model/__init__.py]:
```suggestion
JobStateSummary = NamedTuple("JobStateSummary", [(value, int) for value in enum_values(Job.states)] + [("all_jobs", int)])  # type: ignore[misc]  # Ref https://github.com/python/mypy/issues/848#issuecomment-255237167
```

Still dynamic, so not testable with mypy, but a bit more explicit w.r.t typing.

File: lib/galaxy/model/__init__.py [hunk 3]
```diff
@@ -15,7 +15,10 @@
 import pwd
 import random
 import string
-from collections import defaultdict
+from collections import (
+    defaultdict,
+    namedtuple,
+)
```
Comment (nsoranzo) [inline lib/galaxy/model/__init__.py]:
```suggestion
from collections import defaultdict
```

**Comment by jdavcs** [comment]
@nsoranzo thanks for the suggestions! #15811 builds on top of this one - I'll rebase that one accordingly once this is merged.

**Review by nsoranzo — APPROVED** [review]

**Comment by mvdbeek** [comment]
Thanks @jdavcs!

**Pull request merged**